### ISSUE TITLE
N-dim array with rank zero

### ISSUE THREAD

**Issue by Upabjojr** [sympy/sympy#10510]
A rank zero tensor should be a scalar.

Should N-dim arrays in _sympy.tensor.array_ return the scalar if created with _shape = ()_ ?

There is a bug in the array printer. Rank zero arrays are not correctly displayed.

**Comment by jksuom** [comment]
> A rank zero tensor should be a scalar.

I agree. But I am not quite sure that tensors and arrays should be identified. Perhaps 0-dim arrays should not be allowed at all.

**Comment by Upabjojr** [comment]
> I agree. But I am not quite sure that tensors and arrays should be identified. Perhaps 0-dim arrays should not be allowed at all.

I agree that 0-dim arrays should not be allowed. The problem is to provide an array-like API for scalars. Class methods are not good, because one would need to monkey patch SymPy's core objects.

An example of API supporting scalars (in this case `rank()`) would be:
```
def get_rank(a):
    if isinstance(a, (Matrix, NDimArray)):
        return a.rank()
    return 0
```
Then users would need not to use class methods directly, rather these external functions, to make sure the code won't raise an exception if attempting `a.rank()` or `a.shape` on a scalar.

An ugly alternative is the usage of monkey patching, that is, when you load the array module, some sort of code like this to be executed:
```
Symbol.rank = lambda: 0
Symbol.shape = ()
Expr.rank = lambda: 0
Expr.shape = ()
```
This would force all SymPy objects to have an array-like API.

Otherwise add this API directly to SymPy _Expr_, giving an array-like API to all SymPy objects.

**Comment by jksuom** [comment]
Can you give an example where an array-like API for scalars would be useful?

**Comment by Upabjojr** [comment]
I noticed this problem as a result of total tensor contractions in
https://github.com/sympy/sympy/pull/10511

The issue in that PR has been solved by returning a non-array object if _tensorcontraction_ contracts all axes of the array.

The problem: it hurts the class stability of the result. One may write code that expects results of _tensorcontraction_ to be arrays, and may eventually bump into an exception because of scalars.

**Comment by jksuom** [comment]
I think this is a valid problem. Perhaps it would be possible to introduce a new class for tensors allowing those tensor algebra operations. The name ``Tensor`` is already taken. Unfortunately, it seems to be used for a rather limited class of tensors involving a metric.

A new class having operations like addition, multiplication, contraction, and using arrays as representations might solve the problem. A rank zero tensor would be represented by a scalar. The difference in representation would be invisible to the user.

**Comment by Upabjojr** [comment]
> Perhaps it would be possible to introduce a new class for tensors allowing those tensor algebra operations. The name Tensor is already taken. Unfortunately, it seems to be used for a rather limited class of tensors involving a metric.

I had a thought for a while about a possible class, say _TensorArray_, to represent N-dim arrays with valence markings (e.g. variance/covariance) and possibly also a metric. But it is not necessary, I think that the existing class _Tensor_ could eventually perform that task. _Tensor_ is currently limited to polynomial tensorial expressions and abstract index convention (i.e. indices may not be repeated except for contractions, such restriction makes the tensor well-behaved under coordinate changes).

The problem is that _Tensor_ requires a lot more work to become usable. It can currently associate a _numpy_ N-dim array through the _.data_ property, for some examples:

http://docs.sympy.org/latest/modules/tensor/tensor.html#sympy.tensor.tensor.TensorHead

**Comment by jksuom** [comment]
I am only superficially familiar with the current tensor module, and it seems quite complex with many classes. It would certainly be preferable if you could make that more usable.

BTW, it just occurred to me that it is, in fact, possible to define *sparse* arrays of rank zero.
